### PR TITLE
Added DynamoDB

### PR DESCRIPTION
**Pull request by joshmsamuels**
- Reads emails and courses they are subscribed to from DynamoDB
- Updates last sent timestamp when a notification is sent

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -1,33 +1,60 @@
...
+let ddb = new AWS.DynamoDB.DocumentClient()
+
+// Time between each check in ms (5 minutes)
+const msBetweenChecks = 300000
 
 // Webadvisor course variables
 const courseSemester = 'F20'
 
+// max 5 courses	
+const courses = ['CIS*3260', 'UNIV*2100', 'CIS*1250', 'NUTR*1010']
```
Comment (alexanderMontague) [inline app.js:41]: rebase off your router PR? Or does this need to be in before like the comment below suggests

File: app.js [hunk 2]
```diff
@@ -44,30 +71,40 @@ app.post('/subscribe', [
...
 
-async function start(courses, recipients) {
+async function updateCourseInfo(courses) {
   while (true) {
     try {
       const availableCourseInfo = await checkWebadvisor(courses)
 
-      sendEmail(availableCourseInfo, recipients)
+      let courseKeys = []
+      Object.keys(availableCourseInfo).forEach((courseKey) => {
```
Comment (alexanderMontague) [inline app.js:101]:
```suggestion
      const courseKeys = Object.keys(availableCourseInfo)
```

File: app.js [hunk 3]
```diff
@@ -77,50 +114,60 @@ async function start(courses, recipients) {
...
+    page.waitForNavigation({
+      waitUntil: 'networkidle0'
+    }),
   ])
 
   // Fill out Search for Sections
   await Promise.all([
     page.select('#VAR1', courseSemester),
     ...courses
-      .map((course, i) => {
```
Comment (alexanderMontague) [inline app.js:103]: we should add a linter file so this does not keep happening

File: app.js [hunk 4]
```diff
@@ -192,34 +247,116 @@ async function sendEmail(courseMap, recipients) {
       user: EMAIL_ADDR,
       pass: EMAIL_PWD
     }
-  });  
+  })
+
+  let emailPromises = []
```
Comment (alexanderMontague) [inline app.js:252]:
```suggestion
  const emailPromises = Object.entries(courseCodeEmailMap).map(async ([courseCode, emails], i) => {
          if (!emails || emails.length == 0 || !courseMap[courseCode] || courseMap[courseCode].length == 0) {
            console.info(`no emails to send in batch ${i + 1} -- courses: ${courseMap[courseCode]} -- emails: ${emails}`)
            return
          }
          
          const courseDetailsHTML = formatCourseInfoHTML(courseCode, courseMap[courseCode])

          let courseDetailsHTML = ''
          const response = await transporter.sendMail({
            from: EMAIL_ADDR,
            to: EMAIL_ADDR,
            bcc: emails,
            subject: `Your course selection info for ${courseCode}`,
            html: fs.readFileSync(__dirname + '/templates/coursesAvailableEmail.html', 'utf8').replace('${INSERT_INFO_HERE}', courseDetailsHTML),
           })
           
           .... rest of logic. I wouldnt use push and call the returned closure like it looks like you're doing. The logic is complex tho so i'm banking on everything working lol
        })
```

File: app.js [hunk 5]
```diff
@@ -192,34 +247,116 @@ async function sendEmail(courseMap, recipients) {
...
+ * 
+ * @param {Array} courseCodes - Array of course codes of available courses
+ * @returns {Object} - Keys correspond to the course codes and the value 
+ * is an array of emails to send the course to
+ */
+async function getEmailsCourseCodesFromDyanmo(courseCodes) {
+  let coursesQuery = []
+
+  // Formats all the course codes for DynamoDB
+  courseCodes.forEach(cc => coursesQuery.push({
```
Comment (alexanderMontague) [inline app.js:319]: get out of the habit of defining the array (using let) and then pushing. Defining and mapping is more efficient

```suggestion
  const coursesQuery = courseCodes.map(cc => 'CourseCode': cc)
```

File: app.js [hunk 6]
```diff
@@ -192,34 +247,116 @@ async function sendEmail(courseMap, recipients) {
...
+  try {
+    const responseData = await ddb.batchGet({
+      RequestItems: {
+        Courses: {
+          Keys: coursesQuery
+        }
+      }
+    }).promise()
+
+    responseData.Responses.Courses.forEach((courseData) => {
```
Comment (alexanderMontague) [inline app.js:335]: maybe need optional chaining here? What happens if the request was completed but badly formatted. Not sure the response format but hopefully `.Responses.Courses` is always there and going to be an array. IK with AWS Amplify that is not the case

**Review by alexanderMontague — APPROVED** [review]
👍

File: app.js (hunk 6 above)
Comment (joshmsamuels) [reply app.js:335]: According to the AWS docs it is always there unless there is an error which is why the code is wrapped in a try-catch. 

I’ll do some more testing today though

File: app.js (hunk 1 above)
Comment (joshmsamuels) [reply app.js:41]: This is actually pointing into my router PR. I added it back to get the Dynamo reading/writing working

File: app.js (hunk 2 above)
Comment (joshmsamuels) [reply app.js:101]: What does this change do? Looks like its removing the forEach

File: app.js (hunk 5 above)
Comment (joshmsamuels) [reply app.js:319]: smart!

File: app.js (hunk 3 above)
Comment (joshmsamuels) [reply app.js:103]: Looks like there is a linter file `.prettierrc`

File: app.js (hunk 2 above)
Comment (joshmsamuels) [reply app.js:101]: Ah I see now 🤦

File: app.js (hunk 6 above)
Comment (joshmsamuels) [reply app.js]: Just confirmed that errors are caught by the catch so `Responses.Courses` should always be set

File: app.js (hunk 4 above)
Comment (joshmsamuels) [reply app.js]: I didnt take your change but I basically copied it in my new commit

**Comment by joshmsamuels** [comment]
Closing this to fix my dynamo schema. ill open a new PR soon

**Pull request closed without merging**